### ISSUE TITLE
WebAssembly streaming compilation doesn't work due to missing Content-Type HTTP Header for .wasm files

### ISSUE THREAD

**Issue by michael-andreev** [cockpit-project/cockpit#16099]
Cockpit version: 215-1
OS: Ubuntu 20.04.2 LTS
Page: n/a (new module)

Cockpit doesn't set `Content-Type` HTTP Response Header (MIME type, Media Type) for `.wasm` files (WebAssembly binary modules). For that reason WebAssembly streaming compilation doesn't work and falls back to not optimal "ArrayBuffer instantiation". MIME type should be `application/wasm` in order to allow effective streaming compilation of WebAssembly modules.

Regarding my particular case, I'm getting next warning in Chrome Debug Console from my Framework, when I run my module under Cockpit:
> Streaming compilation failed. Falling back to ArrayBuffer instantiation.  TypeError: Failed to execute 'compile' on 'WebAssembly': Incorrect response MIME type. Expected 'application/wasm'.

Additionally I've tested my module under `nginx/1.21.1` and found that it adds `Content-Type: application/wasm` Header for `.wasm` files by default.

Regarding `application/wasm` MIME (Media) type:
1. `application/wasm` is recommended MIME Type for files containing WebAssembly modules in binary format (`.wasm` files) according WebAssembly Specification ([https://www.w3.org/TR/wasm-core/](https://www.w3.org/TR/wasm-core/)).
2. `application/wasm` MIME type is required for effective WebAssembly streaming compilation according WebAssembly Specification ([https://www.w3.org/TR/wasm-web-api/](https://www.w3.org/TR/wasm-web-api/)). Here's simpler description with examples: [https://developers.google.com/web/updates/2018/04/loading-wasm](https://developers.google.com/web/updates/2018/04/loading-wasm).
3. `application/wasm` is officially registered MIME type ([https://www.iana.org/assignments/media-types/media-types.xhtml](https://www.iana.org/assignments/media-types/media-types.xhtml)).
4. `application/wasm` MIME type already has been added to a very popular NGINX web server ([https://trac.nginx.org/nginx/ticket/1606](https://trac.nginx.org/nginx/ticket/1606)).

### Steps to reproduce

Execute some code with `WebAssembly.instantiateStreaming` in a Cockpit Module, example can be found here: [https://developers.google.com/web/updates/2018/04/loading-wasm](https://developers.google.com/web/updates/2018/04/loading-wasm).

Additionally I could provide a sample Cockpit Module if it would be required.

**Comment by martinpitt** [comment]
Thanks for your report! This shouldn't be too hard to fix, I'm happy to look at this tomorrow. This is fairly easy to reproduce (for a developer) in a built tree:
```
mkdir -p /tmp/cockpit/static/
for i in txt jpg wasm; do echo hello > /tmp/cockpit/static/hello.$i; done
XDG_DATA_DIRS=/tmp G_MESSAGES_DEBUG=all ./cockpit-ws -p 9999 --no-tls
```

Querying:
```
$ curl -I http://localhost:9999/cockpit/static/hello.jpg
HTTP/1.1 200 OK
Access-Control-Allow-Origin: http://localhost:9999
Content-Type: image/jpg

$  curl -I http://localhost:9999/cockpit/static/hello.wasm
HTTP/1.1 200 OK
Access-Control-Allow-Origin: http://localhost:9999
Content-Length: 6
# (no Content-Type at all)
```

**Comment by martinpitt** [comment]
The fix is trivial, in case you want to play around with it:

```diff
--- src/common/cockpitwebresponse.c
+++ src/common/cockpitwebresponse.c
@@ -1796,6 +1796,7 @@ cockpit_web_response_content_type (const gchar *path)
     { ".svg", "image/svg+xml" },
     { ".ttf", "application/octet-stream" }, /* unassigned */
     { ".txt", "text/plain" },
+    { ".wasm", "application/wasm" },
     { ".woff", "application/font-woff" },
     { ".xml", "text/xml" },
   };
```

I'd just like to write a test case for it for a PR.

**Comment by michael-andreev** [comment]
Thank you Martin, it's great! Yeah, I found that row, but unfortunately couldn't build `cockpit` to test it on my Ubuntu Workstation (dependencies issues). So probably will wait for a new version with that fix.

**Comment by martinpitt** [comment]
You can build packages with our CI, in this case `test/image-prepare ubuntu-2004` and grabbing the packages from tmp/build-results/. I built one for you: https://piware.de/tmp/cockpit-ws_248.33.g24275b9ac-1_amd64.deb  With that version it will upgrade automatically once the next official version lands in focal-backports.

**Comment by michael-andreev** [comment]
It's awesome, the `Content-Type` header is presented in your build and now I can use WebAssembly streaming compilation in my module:-)
I also tried your command for CI build. Seems that it builds .DEB package under VM only, while cockpit project should be compiled locally. Unfortunately I have the same issues with dependencies as with running `autogen.sh` script. Probably it's a question for another issue. Anyway thank you for your advice regarding building packages under CI, I'll keep it in mind!

**Comment by martinpitt** [comment]
Right, the webpack is built locally, the VM only builds the C parts. You do need a few dependencies locally (nodejs and git at least, but also autoconf/automake/make). Anyway, glad that it works for you!

Out of interest, is that a public project? I'm curious about how to write a cockpit page in wasm (my first guess would be [rust](https://rustwasm.github.io/)?), do you use cockpit's wire protocol directly for that instead of cockpit.js?

**Comment by michael-andreev** [comment]
Finally I figured out how to install required dependencies on Ubuntu (`sudo apt-get build-dep cockpit` command from your other post helped me). Now I can build packages locally, great.

> Out of interest, is that a public project? I'm curious about how to write a cockpit page in wasm (my first guess would be rust?), do you use cockpit's wire protocol directly for that instead of cockpit.js?

The project that I'm working on is under early development and is not public at this time, but I hope that I'll finish that and release in few months. I decided to try [Blazor](https://www.blazor.net/) framework that runs C# code in WebAssembly. Currently I'm calling cockpit.js from Blazor. Could you tell what you meant about "cockpit wire's protocol"? Is it internal cockpit WebSocket service?

**Comment by martinpitt** [comment]
@michael-andreev : Oh, you can import/call JavaScript from C#? Cool! :sunglasses: 

By "wire protocol" meant the [cockpit websocket JSON protocol](https://github.com/cockpit-project/cockpit/blob/main/doc/protocol.md). It's not really "internal" -- in fact, that is the main thing for which we guarantee stability. cockpit.js is a (fairly thick for some cases) wrapper around that, but in principle you can use all functionality from any other language.

But if you can call cockpit.js, that's certainly easier, so nevermind.

**Comment by michael-andreev** [comment]
Yeah, it's really cool :smile: However WebSocket connection should be much better, since it removes few layers. I definitely should look into that. Thank you for your advice and doc links!
